src/translator/sentence_ranges.h

```diff
@@ -1,52 +1,253 @@
+      flatByteRanges_.emplace_back(sview.data(), sview.size());
+    }
+  };
+
+  /// Adds a sentence, used to load from SentencePiece annotations conveniently.
+  /// encodeWithByteRanges and decodeWithByteRanges.
+  void addSentence(std::vector<string_view_type> &wordRanges) {
+    addSentence(std::begin(wordRanges), std::end(wordRanges));
+  };
 
-  void addSentence(std::vector<string_view> &wordRanges);
-  void addSentence(WordIterator begin, WordIterator end);
+  /// Adds a sentence between two iterators, often useful while constructing
+  /// from parts of a container.
+  void addSentence(WordIterator begin, WordIterator end) {
+    size_t size = flatByteRanges_.size();
+    flatByteRanges_.insert(std::end(flatByteRanges_), begin, end);
+    sentenceBeginIds_.push_back(size);
+  };
 
+  /// clears internal containers, finds use in at least copy-assignment.
   void clear() {
     flatByteRanges_.clear();
     sentenceBeginIds_.clear();
   }
 
+  /// Returns the number of sentences annotated in a blob.
   size_t numSentences() const { return sentenceBeginIds_.size(); }
 
-  // Returns a string_view into the ith sentence.
-  string_view sentence(size_t index) const;
+  /// Returns indices of terminal (word) string_views of a sentence, in
+  /// flatByteRanges_.
+  std::pair<size_t, size_t> sentenceTerminals(size_t sentenceIdx) const {
+    size_t bosId, eosId;
+    bosId = sentenceBeginIds_[sentenceIdx];
+    eosId = sentenceIdx + 1 < numSentences()
+                ? sentenceBeginIds_[sentenceIdx + 1] - 1
+                : flatByteRanges_.size() - 1;
+    assert(bosId < flatByteRanges_.size());
+    assert(eosId < flatByteRanges_.size());
+    return std::make_pair(bosId, eosId);
+  }
+
+  /// Returns a string_view_type into the sentence corresponding to sentenceIdx
+  string_view_type sentence(size_t sentenceIdx) const {
+    auto terminals = sentenceTerminals(sentenceIdx);
+    string_view_type eos, bos;
+    bos = flatByteRanges_[terminals.first];
+    eos = flatByteRanges_[terminals.second];
+    return sentenceBetween(bos, eos);
+  };
+
+  /// Provides access to the internal word correspoding to wordIdx, for the
+  /// sentence corresponding to sentenceIdx.
+  string_view_type word(size_t sentenceIdx, size_t wordIdx) const {
+    size_t offset = sentenceBeginIds_[sentenceIdx];
+    auto terminals = sentenceTerminals(sentenceIdx);
+    assert(offset + wordIdx <= terminals.second);
+    return flatByteRanges_[offset + wordIdx];
+  }
+
+  /// Allows access to a string_view using flatIdx
+  string_view_type wordFromFlatIdx(size_t flatIdx) const {
+    assert(flatIdx < flatByteRanges_.size());
+    return flatByteRanges_[flatIdx];
+  }
+
+  /// Const reference required for use in the conversion function.
+  const std::vector<string_view_type> &flatByteRanges() const {
+    return flatByteRanges_;
+  }
+
+  /// Const reference required for use in the conversion function.
+  const std::vector<size_t> &sentenceBeginIds() const {
```

Comment:
Document what this function returns.

Reply:
Function used in convertion is removed, marking as resolved.